external-crates/move/crates/move-compiler/src/parser/syntax.rs

```diff
@@ -388,57 +406,50 @@ where
     v
 }
 
-/// Attempts to skip tokens until the end of the item in the comma-separated list (which started
-/// with an already consumed starting token) - looks for a matched ending token or after a comma.
-fn skip_comma_list_item(
+/// Attempts to skip tokens until the end of the item in the list with a given separator (which
+/// started with an already consumed starting token) - looks for a matched ending token or a token
+/// appearing after the separator. This helper function is also used to error advance sequences.
+fn list_error_advance(
     context: &mut Context,
     start_token: Tok,
     end_token: Tok,
+    sep_token: Tok,
+    for_list: bool,
     diag: Diagnostic,
-) -> bool {
+) {
     context.env.add_diag(diag);
-    let mut start_tokens: i32 = 0;
-    let mut stop_parsing = false;
+    let mut depth: i32 = 0; // When we find  another start token, we track how deep we are in them
     loop {
+        if for_list {
+            adjust_token(context.tokens, end_token);
+        }
         if context.at_stop_set() {
-            stop_parsing = true;
             break;
         }
-        if context.tokens.at(Tok::Comma) {
-            context.advance();
-            if start_tokens == 0 {
+        if depth == 0 {
+            if context.tokens.at(end_token) {
                 break;
             }
-            continue;
-        };
-        if context.tokens.at(start_token) {
-            start_tokens += 1;
-            context.advance();
-            continue;
-        };
-        if context.tokens.at(end_token) {
-            if start_tokens == 0 {
+            if context.tokens.at(sep_token) {
+                context.advance();
                 break;
             }
-            start_tokens -= 1;
+        }
+
+        if context.tokens.at(sep_token) {
+            assert!(depth > 0);
             context.advance();
-            continue;
-        };
-        if end_token == Tok::Greater && context.tokens.at(Tok::GreaterGreater) {
-            // this is not very pretty but arguably not uglier than adjust_token
-            if start_tokens == 0 || start_tokens == 1 {
-                adjust_token(context.tokens, end_token);
-                break;
-            }
-            start_tokens -= 2;
+        } else if context.tokens.at(start_token) {
+            depth += 1;
+            context.advance();
+        } else if context.tokens.at(end_token) {
+            assert!(depth > 0);
+            depth -= 1;
+            context.advance();
+        } else {
```

Comment:
I appreciate the break out of separateion of concerns, but I would _really_ like `context.advance()` to be a top-level statement in the loop because it guarantees progress and makes it harder for someone to accidentally fall into an infinite loop while extending this code.

Reply:
Done!